File: rust/crates/corlinman-gateway/src/services/chat_service.rs
//! Real `ChatService` implementation — bridges [`corlinman_gateway_api`]
//! callers to the same [`ChatBackend`] that serves `/v1/chat/completions`.
//!
//! The HTTP route and this service share a backend handle so they cannot drift:
//! whatever the Python reasoning loop does for an HTTP request also holds for
//! a QQ message arriving via the channels crate.
//!
//! Scope for M5: surface `TokenDelta` → `InternalChatEvent::TokenDelta`,
//! `ToolCall` → event (informational — the backend still ack's with the
//! placeholder), `Done` → terminal, `Error` → terminal. `AwaitingApproval`
//! and streaming `Usage` frames are skipped (they land with the approval
//! pipeline in M6+).

use std::sync::Arc;

use async_trait::async_trait;
use corlinman_agent_client::tool_callback::{PlaceholderExecutor, ToolExecutor};
use corlinman_core::FailoverReason;
use corlinman_gateway_api::{
    Attachment as ApiAttachment, AttachmentKind as ApiAttachmentKind, ChannelBinding,
    ChatEventStream, ChatService as ChatServiceTrait, InternalChatError, InternalChatEvent,
    InternalChatRequest, Role as ApiRole, Usage as ApiUsage,
};
use corlinman_proto::v1::{
    client_frame, server_frame, Attachment as PbAttachment, AttachmentKind as PbAttachmentKind,
    ChannelBinding as PbChannelBinding, ChatStart, ClientFrame, Message as PbMessage,
    Role as PbRole,
};
use futures::{stream, StreamExt};
use tokio_util::sync::CancellationToken;

use crate::routes::chat::ChatBackend;

/// Gateway-side service that wraps any [`ChatBackend`] so it can be driven
/// from in-process callers via the [`ChatServiceTrait`].
pub struct ChatService {
    backend: Arc<dyn ChatBackend>,
    tool_executor: Arc<dyn ToolExecutor>,
}

impl ChatService {
    /// Bundle a backend with the default placeholder tool executor (matches
    /// the wiring in [`crate::routes::chat::ChatState::new`]).
    pub fn new(backend: Arc<dyn ChatBackend>) -> Self {
        Self {
            backend,
            tool_executor: Arc::new(PlaceholderExecutor),
        }
    }

    /// Customise the tool executor — used by tests; production bundles the
    /// placeholder exec (same as the HTTP route) so tool_calls keep the
    /// Python loop progressing.
    pub fn with_tool_executor(mut self, exec: Arc<dyn ToolExecutor>) -> Self {
        self.tool_executor = exec;
        self
    }
}

#[async_trait]
impl ChatServiceTrait for ChatService {
    async fn run(&self, req: InternalChatRequest, cancel: CancellationToken) -> ChatEventStream {
        let start = build_chat_start(&req);
        let backend = self.backend.clone();
        let executor = self.tool_executor.clone();

        match backend.start(start).await {
            Ok((tx, rx)) => Box::pin(into_event_stream(rx, tx, executor, cancel)),
            Err(err) => {
                // Single terminal error event, then close.
                let ev = InternalChatEvent::Error(InternalChatError::from(err));
                Box::pin(stream::iter(vec![ev]))
            }
        }
    }
}

/// Fold a `ServerFrame` stream into `InternalChatEvent`s. Stops after the
/// first terminal frame (`Done` / `Error`) or when `cancel` fires.
fn into_event_stream(
    rx: crate::routes::chat::BackendRx,
    tx: tokio::sync::mpsc::Sender<ClientFrame>,
    executor: Arc<dyn ToolExecutor>,
    cancel: CancellationToken,
) -> impl futures::Stream<Item = InternalChatEvent> + Send {
    struct State {
        rx: crate::routes::chat::BackendRx,
        tx: tokio::sync::mpsc::Sender<ClientFrame>,
        executor: Arc<dyn ToolExecutor>,
        cancel: CancellationToken,
        done: bool,
    }

    let state = State {
        rx,
        tx,
        executor,
        cancel,
        done: false,
    };

    stream::unfold(state, |mut s| async move {
        if s.done {
            return None;
        }
        loop {
            if s.cancel.is_cancelled() {
                s.done = true;
                return Some((
                    InternalChatEvent::Error(InternalChatError {
                        reason: FailoverReason::Unknown,
                        message: "cancelled".into(),
                    }),
                    s,
                ));
            }

            let next = tokio::select! {
                biased;
                _ = s.cancel.cancelled() => {
                    s.done = true;
                    return Some((
                        InternalChatEvent::Error(InternalChatError {
                            reason: FailoverReason::Unknown,
                            message: "cancelled".into(),
                        }),
                        s,
                    ));
                }
                frame = s.rx.next() => frame,
            };

            match next {
                Some(Ok(frame)) => match frame.kind {
                    Some(server_frame::Kind::Token(t)) => {
                        return Some((InternalChatEvent::TokenDelta(t.text), s));
                    }
                    Some(server_frame::Kind::ToolCall(tc)) => {
                        // Echo the placeholder result so the Python reasoning
                        // loop advances — matches what the HTTP handler does.
                        if let Ok(result) = s.executor.execute(&tc).await {
                            let _ =
                                s.tx.send(ClientFrame {
                                    kind: Some(client_frame::Kind::ToolResult(result)),
                                })
                                .await;
                        }
                        return Some((
                            InternalChatEvent::ToolCall {
                                plugin: tc.plugin,
                                tool: tc.tool,
                                args_json: tc.args_json.into(),
                            },
                            s,
                        ));
                    }
                    Some(server_frame::Kind::Done(d)) => {
                        s.done = true;
                        let usage = d.usage.map(|u| ApiUsage {
                            prompt_tokens: u.prompt_tokens,
                            completion_tokens: u.completion_tokens,
                            total_tokens: u.total_tokens,
                        });
                        return Some((
                            InternalChatEvent::Done {
                                finish_reason: d.finish_reason,
                                usage,
                            },
                            s,
                        ));
                    }
                    Some(server_frame::Kind::Error(e)) => {
                        s.done = true;
                        return Some((
                            InternalChatEvent::Error(InternalChatError {
                                reason: reason_from_proto(e.reason),
                                message: e.message,
                            }),
                            s,
                        ));
                    }
                    Some(server_frame::Kind::Awaiting(_))
                    | Some(server_frame::Kind::Usage(_))
                    | None => {
                        // Not surfaced in M5 — pull the next frame.
                        continue;
                    }
                },
                Some(Err(err)) => {
                    s.done = true;
                    return Some((InternalChatEvent::Error(InternalChatError::from(err)), s));
                }
                None => {
                    // Stream ended without Done — synthesise one so callers
                    // always see a terminal event.
                    s.done = true;
                    return Some((
                        InternalChatEvent::Done {
                            finish_reason: "stop".into(),
                            usage: None,
                        },
                        s,
                    ));
                }
            }
        }
    })
}

fn build_chat_start(req: &InternalChatRequest) -> ChatStart {
    let messages = req
        .messages
        .iter()
        .map(|m| PbMessage {
            role: role_to_proto(m.role) as i32,
            content: m.content.clone(),
            name: String::new(),
            tool_call_id: String::new(),
            content_json: Default::default(),
        })
        .collect();
    let attachments = req.attachments.iter().map(attachment_to_proto).collect();
    let binding = req.binding.as_ref().map(binding_to_proto);
    ChatStart {
        model: req.model.clone(),
        messages,
        tools_json: Vec::new(),
        session_key: req.session_key.clone(),
        binding,
        placeholders: Default::default(),
        temperature: req.temperature.unwrap_or(0.0),
        max_tokens: req.max_tokens.unwrap_or(0),
        stream: req.stream,
        trace: None,
        provider_config_json: Vec::new(),
        attachments,
    }
}

/// Convert the in-process [`ChannelBinding`] to its protobuf twin. The
/// `session_key` field on the proto side is the pre-derived key so the
/// Python agent doesn't need to re-hash.
fn binding_to_proto(b: &ChannelBinding) -> PbChannelBinding {
    PbChannelBinding {
        channel: b.channel.clone(),
        account: b.account.clone(),
        thread: b.thread.clone(),
        sender: b.sender.clone(),
        session_key: b.session_key(),
    }
}

/// Convert the in-process [`ApiAttachment`] to its protobuf twin. The enum
/// mapping is explicit — silently defaulting to `UNSPECIFIED` would drop
/// multimodal inputs without a trace.
fn attachment_to_proto(a: &ApiAttachment) -> PbAttachment {
    let kind = match a.kind {
        ApiAttachmentKind::Image => PbAttachmentKind::Image,
        ApiAttachmentKind::Audio => PbAttachmentKind::Audio,
        ApiAttachmentKind::Video => PbAttachmentKind::Video,
        ApiAttachmentKind::File => PbAttachmentKind::File,
    };
    PbAttachment {
        kind: kind as i32,
        url: a.url.clone().unwrap_or_default(),
        bytes: a.bytes.clone().unwrap_or_default(),
        mime: a.mime.clone().unwrap_or_default(),
        file_name: a.file_name.clone().unwrap_or_default(),
    }
}

fn role_to_proto(role: ApiRole) -> PbRole {
    match role {
        ApiRole::User => PbRole::User,
        ApiRole::Assistant => PbRole::Assistant,
        ApiRole::System => PbRole::System,
        ApiRole::Tool => PbRole::Tool,
    }
}

// Separate converter here because the one in `routes::chat` is private.
fn reason_from_proto(code: i32) -> FailoverReason {
    match code {
        1 => FailoverReason::Billing,
        2 => FailoverReason::RateLimit,
        3 => FailoverReason::Auth,
        4 => FailoverReason::AuthPermanent,
        5 => FailoverReason::Timeout,
        6 => FailoverReason::ModelNotFound,
        7 => FailoverReason::Format,
        8 => FailoverReason::ContextOverflow,
        9 => FailoverReason::Overloaded,
        10 => FailoverReason::Unknown,
        _ => FailoverReason::Unspecified,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::routes::chat::{BackendRx, ChatBackend as ChatBackendTrait};
    use corlinman_core::CorlinmanError;
    use corlinman_gateway_api::Message as ApiMessage;
    use corlinman_proto::v1::{
        Done as PbDone, ErrorInfo, ServerFrame, TokenDelta, ToolCall as PbToolCall,
    };
    use futures::stream;
    use std::pin::Pin;
    use tokio::sync::{mpsc, Mutex};

    /// Minimal scripted backend mirroring the one in routes::chat::mock but
    /// kept local so we don't depend on `#[cfg(test)]` items across modules.
    #[derive(Default)]
    struct ScriptedBackend {
        frames: Mutex<Vec<ServerFrame>>,
    }

    impl ScriptedBackend {
        fn new(frames: Vec<ServerFrame>) -> Self {
            Self {
                frames: Mutex::new(frames),
            }
        }
    }

    #[async_trait]
    impl ChatBackendTrait for ScriptedBackend {
        async fn start(
            &self,
            _start: ChatStart,
        ) -> Result<(mpsc::Sender<ClientFrame>, BackendRx), CorlinmanError> {
            let (tx, _rx) = mpsc::channel::<ClientFrame>(8);
            let frames: Vec<_> = std::mem::take(&mut *self.frames.lock().await)
                .into_iter()
                .map(Ok)
                .collect();
            let s: Pin<Box<_>> = Box::pin(stream::iter(frames));
            Ok((tx, s))
        }
    }

    fn sample_req() -> InternalChatRequest {
        InternalChatRequest {
            model: "m".into(),
            messages: vec![ApiMessage {
                role: ApiRole::User,
                content: "hi".into(),
            }],
            session_key: "abc".into(),
            stream: true,
            max_tokens: None,
            temperature: None,
            attachments: Vec::new(),
            binding: None,
        }
    }

    #[tokio::test]
    async fn tokens_and_done_are_surfaced() {
        let backend = Arc::new(ScriptedBackend::new(vec![
            ServerFrame {
                kind: Some(server_frame::Kind::Token(TokenDelta {
                    text: "hello ".into(),
                    is_reasoning: false,
                    seq: 0,
                })),
            },
            ServerFrame {
                kind: Some(server_frame::Kind::Token(TokenDelta {
                    text: "world".into(),
                    is_reasoning: false,
                    seq: 1,
                })),
            },
            ServerFrame {
                kind: Some(server_frame::Kind::Done(PbDone {
                    finish_reason: "stop".into(),
                    usage: None,
                    total_tokens_seen: 0,
                    wall_time_ms: 0,
                })),
            },
        ]));
        let svc = ChatService::new(backend);
        let cancel = CancellationToken::new();
        let mut s = svc.run(sample_req(), cancel).await;
        let mut text = String::new();
        let mut saw_done = false;
        while let Some(ev) = s.next().await {
            match ev {
                InternalChatEvent::TokenDelta(t) => text.push_str(&t),
                InternalChatEvent::Done { finish_reason, .. } => {
                    assert_eq!(finish_reason, "stop");
                    saw_done = true;
                }
                other => panic!("unexpected event: {other:?}"),
            }
        }
        assert_eq!(text, "hello world");
        assert!(saw_done);
    }

    #[tokio::test]
    async fn tool_call_is_surfaced_and_acked() {
        let backend = Arc::new(ScriptedBackend::new(vec![
            ServerFrame {
                kind: Some(server_frame::Kind::ToolCall(PbToolCall {
                    call_id: "c1".into(),
                    plugin: "p".into(),
                    tool: "t".into(),
                    args_json: b"{}".to_vec(),
                    seq: 0,
                })),
            },
            ServerFrame {
                kind: Some(server_frame::Kind::Done(PbDone {
                    finish_reason: "tool_calls".into(),
                    usage: None,
                    total_tokens_seen: 0,
                    wall_time_ms: 0,
                })),
            },
        ]));
        let svc = ChatService::new(backend);
        let mut s = svc.run(sample_req(), CancellationToken::new()).await;
        let mut saw_tool = false;
        while let Some(ev) = s.next().await {
            if let InternalChatEvent::ToolCall { plugin, tool, .. } = ev {
                assert_eq!(plugin, "p");
                assert_eq!(tool, "t");
                saw_tool = true;
            }
        }
        assert!(saw_tool);
    }

    #[tokio::test]
    async fn error_frame_terminates_stream() {
        let backend = Arc::new(ScriptedBackend::new(vec![ServerFrame {
            kind: Some(server_frame::Kind::Error(ErrorInfo {
                reason: 9,
                message: "overloaded".into(),
                retryable: true,
                upstream_code: String::new(),
            })),
        }]));
        let svc = ChatService::new(backend);
        let mut s = svc.run(sample_req(), CancellationToken::new()).await;
        let ev = s.next().await.expect("at least one event");
        match ev {
            InternalChatEvent::Error(e) => {
                assert_eq!(e.reason, FailoverReason::Overloaded);
                assert_eq!(e.message, "overloaded");
            }
            other => panic!("expected Error, got {other:?}"),
        }
        assert!(s.next().await.is_none());
    }

    #[test]
    fn binding_propagates_to_chat_start() {
        // When the caller supplies a ChannelBinding, build_chat_start must
        // copy every field verbatim and include the derived session_key.
        let mut req = sample_req();
        let b = ChannelBinding::qq_group(100, 200, 300);
        let expected_key = b.session_key();
        req.binding = Some(b);

        let start = build_chat_start(&req);
        let got = start.binding.expect("binding filled through");
        assert_eq!(got.channel, "qq");
        assert_eq!(got.account, "100");
        assert_eq!(got.thread, "200");
        assert_eq!(got.sender, "300");
        assert_eq!(got.session_key, expected_key);
    }

    #[test]
    fn absent_binding_stays_none() {
        let req = sample_req(); // binding: None
        let start = build_chat_start(&req);
        assert!(start.binding.is_none(), "None in → None out");
    }

    #[tokio::test]
    async fn cancel_aborts_with_error() {
        // Backend that yields once and then stalls forever.
        struct Stalling;
        #[async_trait]
        impl ChatBackendTrait for Stalling {
            async fn start(
                &self,
                _start: ChatStart,
            ) -> Result<(mpsc::Sender<ClientFrame>, BackendRx), CorlinmanError> {
                let (tx, _rx) = mpsc::channel::<ClientFrame>(8);
                let s = stream::pending::<Result<ServerFrame, CorlinmanError>>();
                Ok((tx, Box::pin(s)))
            }
        }
        let svc = ChatService::new(Arc::new(Stalling));
        let cancel = CancellationToken::new();
        let mut s = svc.run(sample_req(), cancel.clone()).await;
        cancel.cancel();
        let ev = tokio::time::timeout(std::time::Duration::from_secs(2), s.next())
            .await
            .expect("cancel unblocks stream")
            .expect("at least one event");
        assert!(matches!(ev, InternalChatEvent::Error(_)));
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/services/chat_service.py
"""Real :class:`ChatService` implementation.

Port of :rust:`corlinman_gateway::services::chat_service`. Bridges
in-process callers (channels, scheduler, admin tasks) to the same chat
backend that serves ``/v1/chat/completions``, so an HTTP request and a
QQ-channel message go through identical reasoning-loop wiring.

The Rust crate factors the backend out as a ``trait ChatBackend``;
the Python equivalent is a :class:`Protocol` (structural typing) so
test backends don't have to inherit anything. The production backend
:class:`GrpcAgentChatBackend` wraps a
:class:`corlinman_grpc.agent_client.AgentClient` — i.e. it dials the
Python agent over gRPC, exactly mirroring how the Rust gateway used to
proxy the HTTP request into the Python plane.

Scope mirrors the Rust M5 surface: ``TokenDelta``,
``ToolCall``, ``Done``, ``Error`` are surfaced as the corresponding
:class:`corlinman_server.gateway_api.InternalChatEvent` variants;
``AwaitingApproval`` and standalone ``Usage`` frames are silently
skipped (they land with the approval pipeline in M6+).
"""

from __future__ import annotations

import asyncio
import logging
from collections.abc import AsyncIterator
from typing import Any, Protocol, runtime_checkable

from corlinman_grpc._generated.corlinman.v1 import (
    agent_pb2,
    common_pb2,
)
from corlinman_grpc.agent_client import (
    AgentClient,
    ChatStream,
    PlaceholderExecutor,
    ToolExecutor,
)
from corlinman_grpc.agent_client.types import FailoverReason as GrpcFailoverReason
from corlinman_server.gateway_api import (
    Attachment as ApiAttachment,
    AttachmentKind as ApiAttachmentKind,
    ChannelBinding,
    ChatEventStream,
    ChatServiceBase,
    DoneEvent,
    ErrorEvent,
    InternalChatError,
    InternalChatRequest,
    Message as ApiMessage,
    Role as ApiRole,
    TokenDeltaEvent,
    ToolCallEvent,
    Usage as ApiUsage,
)

__all__ = [
    "ChatBackend",
    "ChatService",
    "GrpcAgentChatBackend",
]


log = logging.getLogger(__name__)


# ─── Backend protocol ────────────────────────────────────────────────


@runtime_checkable
class ChatBackend(Protocol):
    """Structural surface mirroring the Rust ``trait ChatBackend``.

    ``start`` opens an in-process pipeline that the :class:`ChatService`
    drives — the returned ``(tx, rx)`` pair is the same shape as the
    Rust ``(mpsc::Sender<ClientFrame>, BackendRx)`` tuple:

    * ``tx`` — outbound :class:`agent_pb2.ClientFrame` channel
      (``ToolResult`` / ``ApprovalDecision`` / ``Cancel``). Production
      backends forward this to a gRPC ``Agent.Chat`` bidi stream;
      tests can wire a no-op queue.
    * ``rx`` — async iterator of :class:`agent_pb2.ServerFrame` (or
      raised exception) that the service folds into
      :class:`InternalChatEvent` variants.

    The protocol is intentionally minimal — every concrete backend
    (gRPC, scripted-mock, future websocket bridge) implements the same
    two-half pattern.
    """

    async def start(
        self,
        start: agent_pb2.ChatStart,
    ) -> tuple[asyncio.Queue[Any], AsyncIterator[agent_pb2.ServerFrame]]: ...


# ─── ChatService ──────────────────────────────────────────────────────


class ChatService(ChatServiceBase):
    """Gateway-side service that wraps any :class:`ChatBackend` so it
    can be driven from in-process callers via the
    :class:`corlinman_server.gateway_api.ChatService` protocol.

    Mirrors :rust:`corlinman_gateway::services::chat_service::ChatService`:
    holds an :class:`Arc<dyn ChatBackend>` (Python: a shared backend
    reference) plus a default :class:`ToolExecutor` that ack's tool
    calls so the reasoning loop keeps progressing.
    """

    def __init__(
        self,
        backend: ChatBackend,
        *,
        tool_executor: ToolExecutor | None = None,
    ) -> None:
        self._backend = backend
        self._tool_executor: ToolExecutor = tool_executor or PlaceholderExecutor()

    def with_tool_executor(self, executor: ToolExecutor) -> ChatService:
        """Customise the tool executor — used by tests; production
        bundles the placeholder exec (same as the HTTP route) so
        ``tool_calls`` keep the Python loop progressing. Returns
        ``self`` so callers can chain (mirrors the Rust builder shape)."""
        self._tool_executor = executor
        return self

    def run(
        self,
        req: InternalChatRequest,
        cancel: asyncio.Event,
    ) -> ChatEventStream:
        """Open the backend pipeline and yield
        :class:`InternalChatEvent` until the stream terminates.

        Implements the :class:`~corlinman_server.gateway_api.ChatService`
        protocol contract: emits any number of
        :class:`TokenDeltaEvent` / :class:`ToolCallEvent` followed by
        exactly one terminal :class:`DoneEvent` or :class:`ErrorEvent`.
        Honours ``cancel`` between every yield.
        """
        return _run_chat(self._backend, self._tool_executor, req, cancel)


async def _run_chat(
    backend: ChatBackend,
    executor: ToolExecutor,
    req: InternalChatRequest,
    cancel: asyncio.Event,
) -> AsyncIterator[Any]:
    """Async generator implementing the Rust ``into_event_stream`` loop.

    Returned by :meth:`ChatService.run`; callers ``async for ev in s``
    just as Rust callers ``while let Some(ev) = s.next().await``.
    """
    start = _build_chat_start(req)
    try:
        tx, rx = await backend.start(start)
    except Exception as err:  # noqa: BLE001 — surface as terminal error
        yield ErrorEvent(error=_internal_error_from_exception(err))
        return

    # Bridge cancel → drop upstream call. We poll-and-select using
    # ``asyncio.wait`` so a fired cancel unblocks the loop even when
    # the backend has nothing pending.
    cancel_task = asyncio.create_task(cancel.wait())
    try:
        while True:
            if cancel.is_set():
                yield ErrorEvent(
                    error=InternalChatError(reason="unknown", message="cancelled"),
                )
                return

            next_task = asyncio.create_task(_next_frame(rx))
            done, _pending = await asyncio.wait(
                {next_task, cancel_task},
                return_when=asyncio.FIRST_COMPLETED,
            )

            if cancel_task in done:
                next_task.cancel()
                # Drain the cancellation so it doesn't leak as a warning.
                with _suppress_cancelled():
                    await next_task
                yield ErrorEvent(
                    error=InternalChatError(reason="unknown", message="cancelled"),
                )
                return

            try:
                frame = await next_task
            except Exception as err:  # noqa: BLE001 — terminal
                yield ErrorEvent(error=_internal_error_from_exception(err))
                return

            if frame is None:
                # Stream ended without ``Done`` — synthesise one so
                # callers always see a terminal event. Matches the
                # Rust ``None`` arm.
                yield DoneEvent(finish_reason="stop", usage=None)
                return

            kind = frame.WhichOneof("kind")
            if kind == "token":
                yield TokenDeltaEvent(text=frame.token.text)
                continue

            if kind == "tool_call":
                tc = frame.tool_call
                # Echo the placeholder result so the Python reasoning
                # loop advances — matches what the HTTP handler does.
                try:
                    result = await executor.execute(tc)
                    await tx.put(agent_pb2.ClientFrame(tool_result=result))
                except Exception as exc:  # noqa: BLE001 — ack failure is non-fatal
                    log.debug(
                        "chat_service.tool_ack_failed plugin=%s tool=%s err=%s",
                        tc.plugin,
                        tc.tool,
                        exc,
                    )
                yield ToolCallEvent(
                    plugin=tc.plugin,
                    tool=tc.tool,
                    args_json=bytes(tc.args_json),
                )
                continue

            if kind == "done":
                d = frame.done
                usage: ApiUsage | None = None
                if d.HasField("usage"):
                    usage = ApiUsage(
                        prompt_tokens=int(d.usage.prompt_tokens),
                        completion_tokens=int(d.usage.completion_tokens),
                        total_tokens=int(d.usage.total_tokens),
                    )
                yield DoneEvent(finish_reason=d.finish_reason, usage=usage)
                return

            if kind == "error":
                e = frame.error
                yield ErrorEvent(
                    error=InternalChatError(
                        reason=_reason_from_proto(int(e.reason)),
                        message=e.message,
                    ),
                )
                return

            # ``awaiting`` and ``usage`` are not surfaced in this milestone
            # — pull the next frame. ``None`` (unset oneof) is treated
            # the same way.
            continue
    finally:
        cancel_task.cancel()
        with _suppress_cancelled():
            await cancel_task


# ─── Helpers (proto translation) ──────────────────────────────────────


def _build_chat_start(req: InternalChatRequest) -> agent_pb2.ChatStart:
    """Build the protobuf ``ChatStart`` from an
    :class:`InternalChatRequest`. Mirrors :rust:`build_chat_start`."""
    messages = [
        common_pb2.Message(
            role=_role_to_proto(m.role),
            content=m.content,
            name="",
            tool_call_id="",
        )
        for m in req.messages
    ]
    attachments = [_attachment_to_proto(a) for a in req.attachments]
    binding = _binding_to_proto(req.binding) if req.binding is not None else None

    start = agent_pb2.ChatStart(
        model=req.model,
        messages=messages,
        tools_json=b"",
        session_key=req.session_key,
        temperature=float(req.temperature or 0.0),
        max_tokens=int(req.max_tokens or 0),
        stream=req.stream,
        provider_config_json=b"",
        attachments=attachments,
    )
    if binding is not None:
        start.binding.CopyFrom(binding)
    return start


def _binding_to_proto(b: ChannelBinding) -> common_pb2.ChannelBinding:
    """Convert the in-process :class:`ChannelBinding` to its protobuf
    twin. The ``session_key`` field on the proto side is the pre-derived
    key so the Python agent doesn't need to re-hash. Mirrors
    :rust:`binding_to_proto`."""
    return common_pb2.ChannelBinding(
        channel=b.channel,
        account=b.account,
        thread=b.thread,
        sender=b.sender,
        session_key=b.session_key(),
    )


def _attachment_to_proto(a: ApiAttachment) -> agent_pb2.Attachment:
    """Convert :class:`ApiAttachment` → protobuf ``Attachment``. The
    enum mapping is explicit — silently defaulting to ``UNSPECIFIED``
    would drop multimodal inputs without a trace. Mirrors
    :rust:`attachment_to_proto`."""
    if a.kind == ApiAttachmentKind.IMAGE:
        kind = agent_pb2.ATTACHMENT_KIND_IMAGE
    elif a.kind == ApiAttachmentKind.AUDIO:
        kind = agent_pb2.ATTACHMENT_KIND_AUDIO
    elif a.kind == ApiAttachmentKind.VIDEO:
        kind = agent_pb2.ATTACHMENT_KIND_VIDEO
    elif a.kind == ApiAttachmentKind.FILE:
        kind = agent_pb2.ATTACHMENT_KIND_FILE
    else:  # pragma: no cover — exhaustive over StrEnum
        kind = agent_pb2.ATTACHMENT_KIND_UNSPECIFIED
    return agent_pb2.Attachment(
        kind=kind,
        url=a.url or "",
        bytes=a.bytes_ or b"",
        mime=a.mime or "",
        file_name=a.file_name or "",
    )


def _role_to_proto(role: ApiRole) -> common_pb2.Role.ValueType:
    if role == ApiRole.USER:
        return common_pb2.USER
    if role == ApiRole.ASSISTANT:
        return common_pb2.ASSISTANT
    if role == ApiRole.SYSTEM:
        return common_pb2.SYSTEM
    if role == ApiRole.TOOL:
        return common_pb2.TOOL
    return common_pb2.ROLE_UNSPECIFIED  # pragma: no cover


# Lowercase string discriminants matching ``InternalChatError.reason``.
# Same set as ``corlinman_grpc.agent_client.types.FailoverReason``.
_REASON_FROM_PROTO: dict[int, str] = {
    int(GrpcFailoverReason.UNSPECIFIED): "unspecified",
    int(GrpcFailoverReason.BILLING): "billing",
    int(GrpcFailoverReason.RATE_LIMIT): "rate_limit",
    int(GrpcFailoverReason.AUTH): "auth",
    int(GrpcFailoverReason.AUTH_PERMANENT): "auth_permanent",
    int(GrpcFailoverReason.TIMEOUT): "timeout",
    int(GrpcFailoverReason.MODEL_NOT_FOUND): "model_not_found",
    int(GrpcFailoverReason.FORMAT): "format",
    int(GrpcFailoverReason.CONTEXT_OVERFLOW): "context_overflow",
    int(GrpcFailoverReason.OVERLOADED): "overloaded",
    int(GrpcFailoverReason.UNKNOWN): "unknown",
}


def _reason_from_proto(code: int) -> str:
    """Mirror :rust:`reason_from_proto` — unknown codes fall back to
    ``"unspecified"`` so a future proto enum addition doesn't crash
    the event stream."""
    return _REASON_FROM_PROTO.get(code, "unspecified")


def _internal_error_from_exception(exc: BaseException) -> InternalChatError:
    """Lift a connector/transport exception to
    :class:`InternalChatError`. Mirrors the Rust
    ``InternalChatError::from(CorlinmanError)`` blanket impl."""
    reason = getattr(exc, "reason", None)
    if isinstance(reason, str) and reason:
        return InternalChatError(reason=reason, message=str(exc))
    return InternalChatError(reason="unknown", message=str(exc))


# ─── Frame helpers ────────────────────────────────────────────────────


async def _next_frame(
    rx: AsyncIterator[agent_pb2.ServerFrame],
) -> agent_pb2.ServerFrame | None:
    """Drain the next frame from an async iterator, returning ``None``
    on clean end-of-stream so the caller can synthesise a terminal
    ``Done`` event (mirrors the Rust ``Option<...>`` shape)."""
    try:
        return await rx.__anext__()
    except StopAsyncIteration:
        return None


class _suppress_cancelled:
    """Tiny ctx mgr to swallow ``asyncio.CancelledError`` raised by
    awaiting a cancelled task. Equivalent of
    ``contextlib.suppress(asyncio.CancelledError)`` but with the
    explicit naming the chat-service flow expects."""

    def __enter__(self) -> None:  # pragma: no cover — trivial
        return None

    def __exit__(self, _exc_type, exc, _tb) -> bool:  # noqa: ANN001
        return isinstance(exc, asyncio.CancelledError)


# ─── gRPC-backed production backend ──────────────────────────────────


class GrpcAgentChatBackend:
    """Production :class:`ChatBackend` that dials the Python agent over
    ``grpc.aio``.

    Wraps an :class:`AgentClient`. Each :meth:`start` call opens a new
    bidi ``Agent.Chat`` stream and sends ``ChatStart`` as the first
    frame; the returned ``(tx, rx)`` pair is the same shape the
    :class:`ChatService` consumer expects.

    The ``tx`` queue is the same bounded queue the underlying
    :class:`ChatStream` uses internally — see
    :data:`corlinman_grpc.agent_client.CHANNEL_CAPACITY`.
    """

    def __init__(self, client: AgentClient) -> None:
        self._client = client

    async def start(
        self,
        start: agent_pb2.ChatStart,
    ) -> tuple[asyncio.Queue[Any], AsyncIterator[agent_pb2.ServerFrame]]:
        stream: ChatStream = await self._client.chat()
        # First frame must be ``ChatStart`` (cf. Rust ``ChatBackend::start``).
        await stream.send(agent_pb2.ClientFrame(start=start))
        # Hand callers the same internal queue the stream writes into so
        # ``tool_result`` / ``cancel`` frames flow back into the bidi
        # half-channel without an extra queue layer.
        tx: asyncio.Queue[Any] = stream._tx  # noqa: SLF001 — same-package access
        return tx, _ServerFrameIter(stream)


class _ServerFrameIter:
    """Async iterator wrapper around :class:`ChatStream` that yields
    raw protobuf frames (the inner half-channel reads them directly
    via ``grpc.aio``'s ``__aiter__``)."""

    def __init__(self, stream: ChatStream) -> None:
        self._stream = stream
        self._aiter = stream.__aiter__()

    def __aiter__(self) -> _ServerFrameIter:
        return self

    async def __anext__(self) -> agent_pb2.ServerFrame:
        return await self._aiter.__anext__()
